src/client.rs

```diff
@@ -0,0 +1,255 @@
+use std::{collections::HashMap, sync::Arc};
+
+use futures_util::TryFutureExt;
+use protobuf::Message;
+use tokio::sync::Mutex;
+
+use crate::{
+    protocol::{
+        index::{
+            CreatePort, CreatePortResponse, Request, RequestModule, RequestModuleResponse,
+            Response, RpcMessageTypes,
+        },
+        parse::{build_message_identifier, parse_protocol_message},
+    },
+    transports::{Transport, TransportEvent},
+};
+
+pub type ClientResult<T> = Result<T, ClientError>;
+
+#[derive(Debug)]
+pub enum ClientError {
+    ProtocolError,
+    TransportError,
+    TransportNotAttached,
+    PortCreationError,
+    ProcedureNotFound,
+    UnknownError,
+}
+
+pub struct RpcClient {
+    ports: HashMap<String, RpcClientPort>,
+    client_request_dispatcher: Arc<ClientRequestDispatcher>,
+}
+
+impl RpcClient {
+    pub async fn new<T: Transport + Send + Sync + 'static>(transport: T) -> ClientResult<Self> {
+        println!("Client IsConnected: {}", transport.is_connected());
+        let transport = Self::establish_connection(transport)
+            .map_err(|err| err)
+            .await?;
+        let transport = Arc::new(transport);
+
+        println!("Client IsConnected: {}", transport.is_connected());
+
+        let client = Self {
+            ports: HashMap::new(),
+            client_request_dispatcher: Arc::new(ClientRequestDispatcher::new(transport)),
+        };
+        Ok(client)
+    }
+
+    async fn establish_connection<T: Transport + Send + Sync + 'static>(
+        mut transport: T,
+    ) -> ClientResult<T> {
+        // Send empty message to notify connection
+        transport
+            .send(vec![0])
+            .await
+            .map_err(|_| ClientError::TransportError)?;
+
+        match transport.receive().await {
+            Ok(response) => {
+                if let TransportEvent::Connect = response {
```

Comment:
maybe we can be more specific on the pattern matching and do something like:

 ```suggestion
            Ok(TransportEvent::Connect) => {
```

Reply:
Cool, didn't know this syntax!